### PR TITLE
Pass errorHandler from UserProfile to AddonsByAuthorsCards

### LINKED ISSUE #5442: Navigating to a non existent page in More add-on by author cards in the User profile page will lock the cards in a loading state
STR:

1. Load AMO -dev
2. Open a user profile page that lists both add-ons and themes - see https://addons-dev.allizom.org/en-US/firefox/user/AlexandraMoga1/?page_e=7&page_t=6
3. Set the page query for the two cards to a non-existent page (i.e. 7 for add-ons and 6 for themes for this case)

Actual result:
The two cards are locked in a loading state when a non-existent page is requested

Expected result:
Display (maybe) a Page not found on the card that is causing the error to appear

Notes:
- reproduced on AMO -dev with FF60, WIn10x64

[1] Page when both cards are requesting an invalid page
screen shot 2018-06-28 at 09 11 21-fullpage

[2] Page when only on card is requesting an invalid page

image

### PR DESCRIPTION
**Pull request by willdurand**
Fix #5442

---

This PR allows to return a 404 when extensions and/or themes on a user
profile cannot be loaded (and the API returns a 404 when trying to fetch
these add-ons), likely due to an invalid query parameter value.

Passing the `errorHandler` from the `UserProfile` component to the
`AddonsByAuthorsCard` components is required to make sure we return a
`NotFound` component on the server, thus returning a real `404` HTTP
response.

I also decided to hide the `AddonsByAuthorsCard` when an error occurs,
because it is better than being stuck in a loading state. On an add-on
detail page, it is "secondary data" so does not matter much I guess,
and on the user profile, the `UserProfile` will display the error.

I don't know if the added tests are enough or if we could find a better
patch, but I don't really see how we could do better here.

### CODE REVIEW

File: src/amo/components/AddonsByAuthorsCard/index.js [hunk 1]
```diff
@@ -205,6 +209,13 @@ export class AddonsByAuthorsCardBase extends React.Component<InternalProps> {
       type,
     } = this.props;
 
+    invariant(errorHandler, 'errorHandler is required');
+
+    if (errorHandler.hasError()) {
```
Comment (kumar303) [inline src/amo/components/AddonsByAuthorsCard/index.js]: This seems like the wrong thing to do because it is swallowing any error that may have been produced by the sagas in this component. If I remove this if statement, the original bug you set out to fix is still fixed: I see the 404 page.

However, swallowing errors looks like an unrelated bug that you may as well fix. I think you can fix it like this:

```diff
diff --git a/src/amo/components/AddonsByAuthorsCard/index.js b/src/amo/components/AddonsByAuthorsCard/index.js
index e9c395e43..74873c400 100644
--- a/src/amo/components/AddonsByAuthorsCard/index.js
+++ b/src/amo/components/AddonsByAuthorsCard/index.js
@@ -211,11 +211,6 @@ export class AddonsByAuthorsCardBase extends React.Component<InternalProps> {
 
     invariant(errorHandler, 'errorHandler is required');
 
-    if (errorHandler.hasError()) {
-      log.warn('Captured API Error:', errorHandler.capturedError);
-      return null;
-    }
-
     if (!loading && (!addons || !addons.length)) {
       return null;
     }
@@ -341,17 +336,20 @@ export class AddonsByAuthorsCardBase extends React.Component<InternalProps> {
     }
 
     return (
-      <AddonsCard
-        addons={addons}
-        className={classnames}
-        footer={paginator}
-        header={header}
-        loading={loading}
-        placeholderCount={numberOfAddons}
-        showMetadata
-        showSummary={showSummary}
-        type={type}
-      />
+      <React.Fragment>
+        {errorHandler.renderErrorIfPresent()}
+        <AddonsCard
+          addons={addons}
+          className={classnames}
+          footer={paginator}
+          header={header}
+          loading={loading}
+          placeholderCount={numberOfAddons}
+          showMetadata
+          showSummary={showSummary}
+          type={type}
+        />
+      </React.Fragment>
     );
   }
 }

```

File: tests/unit/amo/components/TestAddonsByAuthorsCard.js [hunk 2]
```diff
@@ -968,4 +970,31 @@ describe(__filename, () => {
       );
     });
   });
+
+  it('does not render anything when an API error is thrown', () => {
```
Comment (kumar303) [inline tests/unit/amo/components/TestAddonsByAuthorsCard.js]: Why is this needed? It doesn't seem necessary to solve the bug of rendering a 404 page.

File: tests/unit/amo/components/TestUserProfile.js [hunk 3]
```diff
@@ -896,6 +919,62 @@ describe(__filename, () => {
...
+      id: 'some-id',
+      dispatch: store.dispatch,
+    });
+
+    const root = renderUserProfile({ errorHandler, store });
+
+    // Simulate an error thrown inside the second AddonsByAuthorsCard.
+    root
+      .find(AddonsByAuthorsCard)
+      .at(1)
```
Comment (kumar303) [inline tests/unit/amo/components/TestUserProfile.js]: When you have two tests that look the same, it's an indication that you aren't testing the unit of logic in the correct place.

I think you can replace these two tests with one test that makes sure `UserProfile` renders a 404 `NotFound` component given an `errorHandler` with a 404 error. It doesn't matter *where* the error handler gets its error from. The set up could look like:

```js
const { store } = dispatchSignInActions();
const errorHandler = new ErrorHandler({
  id: 'some-id',
  dispatch: store.dispatch,
});
errorHandler.handle(
  createApiError({
    response: { status: 500 },
    apiURL: 'https://some/api/endpoint',
    jsonResponse: { message: 'internal server error' },
  })
);

const root = renderUserProfile({ errorHandler, store });
// Check for the error...
```

The reason you don't need to test the `errorComponent` of each `AddonsByAuthorsCard` is because you already have a test to make sure that their `errorComponent` props are coming from `UserProfile`.

File: src/amo/components/AddonsByAuthorsCard/index.js (hunk 1 above)
Comment (kumar303) [reply src/amo/components/AddonsByAuthorsCard/index.js]: > I also decided to hide the AddonsByAuthorsCard when an error occurs,
because it is better than being stuck in a loading state.

Ah, I just read this part in your description.

In this case, you should switch it in an if statement, like:

```js
render() {
  if (errorHandler.hasError()) {
    return errorHandler.renderHoweverItsCalled();
  } else {
    return <AddonsCard ...>
  }
}
```
I don't think you should hide the error, though.

File: src/amo/components/UserProfile/index.js [hunk 4]
```diff
@@ -385,13 +383,14 @@ export class UserProfileBase extends React.Component<InternalProps> {
...
+          {params &&
+            params.username && (
               <div className="UserProfile-addons-and-reviews">
                 <AddonsByAuthorsCard
                   addonType={ADDON_TYPE_EXTENSION}
-                  authorDisplayName={user.name}
-                  authorUsernames={[user.username]}
+                  authorDisplayName={user ? user.name : params.username}
+                  authorUsernames={[params.username]}
+                  errorHandler={errorHandler}
```
[the remark this answers is not in the document]
Comment (willdurand) [reply src/amo/components/UserProfile/index.js]: Yeah, definitely. It is fine here but nothing we can generalize :/

File: src/amo/components/AddonsByAuthorsCard/index.js (hunk 1 above)
Comment (willdurand) [reply src/amo/components/AddonsByAuthorsCard/index.js]: >If I remove this if statement, the original bug you set out to fix is still fixed: I see the 404 page.

Yes, I added this after..

File: tests/unit/amo/components/TestUserProfile.js (hunk 3 above)
Comment (willdurand) [reply tests/unit/amo/components/TestUserProfile.js]: right, I don't know what I was thinking.

**Comment by willdurand** [comment]
So I believe I addressed all the comments.

File: tests/unit/amo/components/TestUserProfile.js (hunk 3 above)
Comment (kumar303) [reply tests/unit/amo/components/TestUserProfile.js]: Heh, no worries. It's always easier to spot when doing a code review.
